t3f/ops.py

```diff
@@ -731,63 +731,161 @@ def add(tt_a, tt_b):
   Args:
-    tt_left: `TensorTrain`, TT-tensor or TT-matrix
-    right: `TensorTrain`, TT-tensor or TT-matrix, OR a number.
+    tt_left: `TensorTrain` OR `TensorTrainBatch`
+    right: `TensorTrain` OR `TensorTrainBatch` OR a number.
 
   Returns
-    a `TensorTrain` object corresponding to the element-wise product of the
-    arguments.
+    a `TensorTrain` or `TensorTrainBatch` object corresponding to the
+    element-wise product of the arguments.
 
   Raises
-    ValueError if the arguments shapes do not coincide.
+    ValueError if the arguments shapes do not coincide or broadcasting is not
+    possible.
   """
+
+  is_left_batch = isinstance(tt_left, TensorTrainBatch)
+  is_right_batch = isinstance(right, TensorTrainBatch)
+
+  is_batch_case = is_left_batch or is_right_batch
+
   if not isinstance(right, TensorTrainBase):
     # Assume right is a number, not TensorTrain.
     tt_cores = list(tt_left.tt_cores)
     tt_cores[0] = right * tt_cores[0]
     out_ranks = tt_left.get_tt_ranks()
+    if (isinstance(tt_left, TensorTrainBatch)):
+        out_batch_size = tt_left.batch_size
   else:
     ndims = tt_left.ndims()
+
     if tt_left.is_tt_matrix() != right.is_tt_matrix():
       raise ValueError('The arguments should be both TT-tensors or both '
                        'TT-matrices')
 
-    if tt_left.get_shape() != right.get_shape():
+    if tt_left.get_raw_shape() != right.get_raw_shape():
       raise ValueError('The arguments should have the same shape.')
 
+    out_batch_size = 1
+    dependencies = []
+    cant_determine_if_broadcast = False
+    if is_left_batch and is_right_batch:
+      if tt_left.batch_size is None and right.batch_size is None:
+        cant_determine_if_broadcast = True
+      elif tt_left.batch_size is None and right.batch_size is not None:
+        if right.batch_size > 1:
+            cant_determine_if_broadcast = True
+      elif tt_left.batch_size is not None and right.batch_size is None:
+        if tt_left.batch_size > 1:
+            cant_determine_if_broadcast = True
+
+    if cant_determine_if_broadcast:
+      # Cannot determine if broadcasting is needed. Avoid broadcasting and
+      # assume elementwise multiplication AND add execution time assert to print
+      # a better error message if the batch sizes turn out to be different.
+
+      message = ('The batch sizes were unknown on compilation stage, so '
+                 'assumed elementwise multiplication (i.e. no broadcasting). '
+                 'Now it seems that they are different after all :')
+
+      data = [message, shapes.lazy_shape(tt_left)[0], ' x ',
+              shapes.lazy_shape(right)[0]]
+      bs_eq = tf.assert_equal(shapes.lazy_batch_size(tt_left),
+                              shapes.lazy_batch_size(right), data=data)
+
+      dependencies.append(bs_eq)
+
+    do_broadcast = shapes.is_batch_broadcasting_possible(tt_left, right)
+    if not do_broadcast and not cant_determine_if_broadcast:
```

Comment:
Don't get this if, why if broadcasting not possible and we can determine broadcasting it should raise error?